### PR TITLE
Refactors port types completely and adds the option type. Refactors options to use this new type

### PR DESCRIPTION
**Pull request by Watermelon914**
## About The Pull Request

Refactors port types to have global datum handlers instead of hardcoded behaviours.
Adds the option type that has replaced the current option system.

Additionally, string types will now accept any types and will implicitly convert them to string. For atoms, this means they'll still appear as "entity" unless they use the tostring type.

Preview of the options type:
image

String output ports can be connected to these option ports, however, if an invalid input is passed, it will automatically take the first option by default.

## Why It's Good For The Game

Better handling of types because previously the behaviour was just hardcoded in which was really shitty. Gives way for better and more interesting datatypes.

## Changelog
:cl:
refactor: Refactors datatype code completely for integrated circuits.
refactor: Options have been moved to their own input port with their own special type known as the "Option" type.
/:cl:

### CODE REVIEW

File: code/__DEFINES/wiremod.dm [hunk 1]
```diff
@@ -4,6 +4,9 @@
 /// Define to automatically handle calling the output port. Will not call the output port if the input_received proc returns TRUE.
 #define TRIGGER_CIRCUIT_COMPONENT(component, port) if(!component.input_received(port) && (component.circuit_flags & CIRCUIT_FLAG_OUTPUT_SIGNAL)) component.trigger_output.set_output(COMPONENT_SIGNAL)
 
+/// Handles adding option ports more simply
+#define SET_OPTION_LIST(option_list) port_type = /datum/port/input/option, extra_args = list(possible_options = option_list)
```
Comment (Mothblocks) [inline code/__DEFINES/wiremod.dm:8]: I don't really like macros like these, can you just make creating an option its own proc so we don't do anything funky?

File: code/modules/wiremod/component.dm [hunk 2]
```diff
@@ -45,28 +45,20 @@
...
 
-	/// The current selected option
-	var/current_option
-	/// The options that this component can take on. Limited to strings
-	var/list/options
-
 	// Whether the component is removable or not. Only affects user UI
 	var/removable = TRUE
 
+/obj/item/circuit_component/proc/populate_options()
```
Comment (Mothblocks) [inline code/modules/wiremod/component.dm:51]: You lost the documentation.

File: code/modules/wiremod/datatypes/option.dm [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+/datum/circuit_datatype/option
+	datatype = PORT_TYPE_OPTION
+	color = "violet"
+	datatype_flags = DATATYPE_FLAG_ALLOW_MANUAL_INPUT
+
+/datum/circuit_datatype/option/is_compatible(datum/port/gained_port)
+	if(istype(gained_port, /datum/port/input/option))
+		return TRUE
+	return FALSE
```
Comment (Mothblocks) [inline code/modules/wiremod/datatypes/option.dm:17]:
```suggestion
	return istype(gained_port, /datum/port/input/option)
```

File: code/modules/wiremod/integrated_circuit.dm [hunk 4]
```diff
@@ -241,6 +241,7 @@
 				"connected_to" = REF(port.connected_port),
 				"color" = port.color,
 				"current_data" = current_data,
+				"datatype_data" = port.datatype_ui_data(user)
```
Comment (Mothblocks) [inline code/modules/wiremod/integrated_circuit.dm:244]:
```suggestion
				"datatype_data" = port.datatype_ui_data(user),
```

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: code/__DEFINES/wiremod.dm (hunk 1 above)
Comment (Watermelon914) [reply code/__DEFINES/wiremod.dm:8]: sure

File: code/modules/wiremod/component.dm (hunk 2 above)
Comment (Watermelon914) [reply code/modules/wiremod/component.dm:51]: shit

File: code/modules/wiremod/component.dm [hunk 5]
```diff
@@ -48,9 +48,13 @@
 	// Whether the component is removable or not. Only affects user UI
 	var/removable = TRUE
 
+/// Called when the option ports should be set up
 /obj/item/circuit_component/proc/populate_options()
 	return
 
+/obj/item/circuit_component/proc/add_option_port(name, list/list_to_use)
```
Comment (Mothblocks) [inline code/modules/wiremod/component.dm:55]: Document this as well.

File: code/__DEFINES/wiremod.dm [hunk 6]
```diff
@@ -114,6 +116,11 @@
 #define COMP_PROC_GLOBAL "Global"
 #define COMP_PROC_OBJECT "Object"
 
+// Bar overlay component
+#define BAR_OVERLAY_LIMIT 10
```
Comment (Mothblocks) [inline code/__DEFINES/wiremod.dm:120]: Why did you move these out? They're only needed in that one file.

**Review by Mothblocks — CHANGES_REQUESTED** [review]
Last nitpick

File: code/__DEFINES/wiremod.dm (hunk 6 above)
Comment (Watermelon914) [reply code/__DEFINES/wiremod.dm:120]: since I do it for other components, e.g. arrest consoles, so I thought it made sense to move the define out to the wiremod.dm file.

File: code/__DEFINES/wiremod.dm [hunk 7]
```diff
@@ -114,6 +116,10 @@
 #define COMP_PROC_GLOBAL "Global"
 #define COMP_PROC_OBJECT "Object"
 
+// Bar overlay component
+#define COMP_BAR_OVERLAY_VERTICAL "Vertical"
```
Comment (Mothblocks) [inline code/__DEFINES/wiremod.dm:120]: Move these new defines back where they were too.

**Review by Mothblocks — CHANGES_REQUESTED** [review]
Still just need those defines back.
